### PR TITLE
Dimensionality Reducers

### PR DESCRIPTION
**Pull request by Zak7062**
Includes component analyzers and feature selectors in two separate files, along with the corresponding unit tests, which just check to see if the dataframe is smaller after the model has been run.

### CODE REVIEW

File: tests/test_classification_learning.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def test_random_forest(self):
         # Arrange
         data = ft.get_data(setup.get_datasets_path(), "titanic_train.csv", ',')
         x_train, x_test, y_train, y_test = ft.get_train_test(data, "Survived")
-
+        print(y_train)
```
Comment (asclines) [inline tests/test_classification_learning.py]: Is this a leftover debug print?

Comment (Zak7062) [inline tests/test_classification_learning.py]: ....uhh...yes

File: modules/toolbox/framework_tools.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
 from modules.toolbox import *
+import pandas as pd
+from sklearn import preprocessing
+from sklearn import cross_validation
 
 # Helper functions for the whole framework to use
 def get_train_test(pandas_data, target_col):
-    pandas_data = scale_numeric_data(pandas_data)
+    #pandas_data = scale_numeric_data(pandas_data)
```
Comment (ASAAR) [inline modules/toolbox/framework_tools.py]: Just wondering why this is commented out. Any particular reason? Scaling is important for removing bias towards one variable or another

File: reducers/.idea/reducers.iml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="PYTHON_MODULE" version="4">
+  <component name="NewModuleRootManager">
+    <content url="file://$MODULE_DIR$" />
+    <orderEntry type="inheritedJdk" />
```
Comment (ASAAR) [inline reducers/.idea/reducers.iml]: @asclines is this something that should be ignored?

File: reducers/.idea/test_feature_selectors.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+from unittest import TestCase
+from modules.toolbox import framework_tools as ft
+from modules.toolbox import ml_runners as mr
+from reducers import scikit_feature_selectors
+import setup
+
+class TestFeatureSelectors(TestCast):
+
+    def test_VarianceThreshold_selector(self):
```
Comment (ASAAR) [inline reducers/.idea/test_feature_selectors.py]: @Zak7062 this is nit picky, so I'm sorry. But function names, according to Google standards, should be like this "function_name"

File: reducers/scikit_component_analyzers.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+#http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html
+def Principle_Component_Analyzer(data):
+    pca = PCA(n_components=3) #if components is not specified, all components are kept.
+    pca.fit_transform(data) #This both fits pca and transforms in 1 command
+    pca = (pd.DataFrame(pca.components_, columns=data.columns)) #Retain feature labels
+    return pca
+
+#http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html#sklearn.decomposition.FastICA
+def Independent_Component_Analyzer(data):
```
Comment (ASAAR) [inline reducers/scikit_component_analyzers.py]: Function names

File: reducers/scikit_feature_selectors.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+def VarianceThreshold_selector(data,target):
+    x_train, x_test, y_train, y_test = ft.get_train_test(data, target)
+
+    #Select Model
+    selector = VarianceThreshold(0) #Defaults to 0.0, e.g. only remove features with the same value in all samples
+
+    #Fit the Model
+    selector.fit(x_train, y_train)
+    data.drop(target, 1, inplace=True) #Remove target feature
+    features = selector.get_support(indices = True) #returns an array of integers corresponding to nonremoved features
```
Comment (ASAAR) [inline reducers/scikit_feature_selectors.py]: @Zak7062 One thing to keep in mind here is that there seems to be some duplicate code between this function and the next two. If at all possible when writing code, it's best to reduce duplication and separate out functionality that can be separated into a new function.

File: tests/test_feature_selection.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+sys.path.insert(0, os.path.abspath('..'))
+
+from unittest import TestCase
+from modules.toolbox import framework_tools as ft
+from modules.toolbox import ml_runners as mr
+from reducers import scikit_feature_selectors
+import setup
+
+class TestFeatureSelection(TestCase):
+    def test_VarianceThreshold_selector(self):
```
Comment (ASAAR) [inline tests/test_feature_selection.py]: Function names

File: reducers/.idea/reducers.iml (hunk 3 above)
Comment (asclines) [inline reducers/.idea/reducers.iml]: Yes, .idea/ folders should be ignored

File: reducers/.idea/test_feature_selectors.py (hunk 4 above)
Comment (asclines) [inline reducers/.idea/test_feature_selectors.py]: @Zak7062 @RyanMcBerg @ASAAR https://google.github.io/styleguide/pyguide.html

File: modules/toolbox/framework_tools.py (hunk 2 above)
Comment (Zak7062) [inline modules/toolbox/framework_tools.py]: When i don't have it commented out, I get a lot of errors complaining about things being converted.

**Pull request merged**